### ISSUE TITLE
Support regex matching for URLRouter placeholders

### ISSUE THREAD

**Issue by s-ludwig** [rejectedsoftware/vibe.d#1479]
See #1062.

Suggested syntax coming from Python: `router.get("/items/:(?P<var>[0-9]+)/something/", ...)`

**Comment by SkyzohKey** [comment]
Another nice syntax could be `router.get("/items/:var([0-9]+)/something/", ...)`

**Comment by Mihail-K** [comment]
Could this also be extended to add strong parameter types?

ie. `router.get("/items/:var(uint)/something/", ...)`

**Comment by s-ludwig** [comment]
It won't work at that level, because the router takes runtime strings. But the `vibe.web.web`/`vibe.web.rest` modules already provide a type checked interface on top.

**Comment by Mihail-K** [comment]
I actually meant that for baked in regex patterns to match those types, so as to not have to repeat yourself over and over, whenever you want a URL parameter to match an integer, for example.

**Comment by s-ludwig** [comment]
Yes, but since it is based on a runtime parameter, that would mean that it can only work with a set of known types, which either limits its usefulness or makes for a relatively cumbersome interface (need to `register()` types up front). It would also mean that the parsing work has to be done twice (worse performance and risk of having inconsistent parsers).

Generally, I don't think the router should be extended with high-level functionality, but should stay focused on it's main task of performing plain string matching as much as possible. Layering this on top is perfectly possible and avoids the mentioned issues, as the web interface generator shows.

**Comment by Mihail-K** [comment]
Fair point. Adding this sort of functionality in project-space is equally viable.

**Comment by wilzbach** [comment]
A first step could be to add support for optional parameters that behave like this:

```
/foo/:var?
/foo/bar [match with var="bar"]
/foo/bar [match with var=""]
```

or in D code:

```d
@path("/foo/:var?", var)
void getFoo(string var = "defaultValue")
```

At the moment having optional parameters in a `path` with the REST or Web wrapper is a bit non-DRY:

```
void getFoo() {
   getFoo("defaultValue");
}
@path("/foo/:var", var)
void getFoo(string var)
```
